Engine/AI/NavMesh.h

```diff
@@ -0,0 +1,190 @@
+#ifndef _NAVMESH_H_
+#define _NAVMESH_H_
+
+#include "AI/SampleDebugDraw.h"
+#include "AI/DebugDrawGL.h"
+#include "Helper/Timer.h"
+
+#include <MathGeoLib.h>
+#include <pcg_basic.h>
+#include <recast/Detour/DetourNavMesh.h>
+#include <recast/Recast/Recast.h>
+#include <vector>
+
+class PanelNavMesh;
+class dtNavMeshQuery;
+class dtNavMesh;
+class dtMeshTile;
+class DetourNavMesh;
+
+
+static const int MAX_POLYS = 256;
+static const int MAX_SMOOTH = 2048;
+
+
+enum PathMode
+{
+	FOLLOW_PATH,
+	STRAIGHT_PATH
+};
+
+enum DrawMode
+{
+	DRAWMODE_NAVMESH,
+	DRAWMODE_NAVMESH_TRANS,
+	DRAWMODE_NAVMESH_BVTREE,
+	DRAWMODE_NAVMESH_NODES,
+	DRAWMODE_NAVMESH_INVIS,
+	DRAWMODE_MESH,
+	DRAWMODE_VOXELS,
+	DRAWMODE_VOXELS_WALKABLE,
+	DRAWMODE_COMPACT,
+	DRAWMODE_COMPACT_DISTANCE,
+	DRAWMODE_COMPACT_REGIONS,
+	DRAWMODE_REGION_CONNECTIONS,
+	DRAWMODE_RAW_CONTOURS,
+	DRAWMODE_BOTH_CONTOURS,
+	DRAWMODE_CONTOURS,
+	DRAWMODE_POLYMESH,
+	DRAWMODE_POLYMESH_DETAIL,
+	MAX_DRAWMODE
+};
+
+enum SamplePartitionType
+{
+	SAMPLE_PARTITION_WATERSHED,
+	SAMPLE_PARTITION_MONOTONE,
+	SAMPLE_PARTITION_LAYERS,
+};
+
+enum SamplePolyFlags
+{
+	SAMPLE_POLYFLAGS_WALK = 0x01,		// Ability to walk (ground, grass, road)
+	SAMPLE_POLYFLAGS_SWIM = 0x02,		// Ability to swim (water)
+	SAMPLE_POLYFLAGS_DOOR = 0x04,		// Ability to move through doors
+	SAMPLE_POLYFLAGS_JUMP = 0x08,		// Ability to jump
+	SAMPLE_POLYFLAGS_DISABLED = 0x10,	// Disabled polygon
+	SAMPLE_POLYFLAGS_ALL = 0xffff		// All abilities
+};
+
+class NavMesh
+{
+public:
+	NavMesh();
+	~NavMesh();
+
+	bool CleanUp();
+	bool Update();
+
+	bool CreateNavMesh();
+	void RenderNavMesh(ComponentCamera& camera);
+	void InitAABB();
+
+	bool FindPath(float3& start, float3& end, std::vector<float3>& path);
+
+	void SaveNavMesh(unsigned char* nav_data, unsigned int nav_data_size) const;
+	void LoadNavMesh();
+
+	inline SampleDebugDraw& GetDebugDraw() { return m_dd; }
+
+private:
+	void GetVerticesScene();
+	void GetIndicesScene();
+	void GetNormalsScene();
+
+	bool GetSteerTarget(dtNavMeshQuery* navQuery, const float* startPos, const float* endPos,
+		const float minTargetDist,
+		const dtPolyRef* path, const int pathSize,
+		float* steerPos, unsigned char& steerPosFlag, dtPolyRef& steerPosRef,
+		float* outPoints = 0, int* outPointCount = 0);
+
+	inline bool InRange(const float* v1, const float* v2, const float r, const float h);
+	int FixUpShortcuts(dtPolyRef* path, int npath, dtNavMeshQuery* nav_query);
+	int FixUpCorridor(dtPolyRef* path, const int npath, const int max_path,
+		const dtPolyRef* visited, const int nvisited);
+
+	float DistancePtLine2d(const float* pt, const float* p, const float* q) const
+	{
+		float pqx = q[0] - p[0];
+		float pqz = q[2] - p[2];
+		float dx = pt[0] - p[0];
+		float dz = pt[2] - p[2];
+		float d = pqx * pqx + pqz * pqz;
+		float t = pqx * dx + pqz * dz;
+		if (d != 0) t /= d;
+		dx = p[0] + t * pqx - pt[0];
+		dz = p[2] + t * pqz - pt[2];
+		return dx * dx + dz * dz;
+	}
```

Comment:
If this definition it's not strictly necessary, this function should be defined inside `.cpp` file.